vector/src/main/java/im/vector/app/features/roomprofile/settings/joinrule/RoomJoinRuleActivity.kt

```diff
@@ -0,0 +1,149 @@
+import im.vector.app.core.extensions.commitTransaction
+import im.vector.app.core.extensions.toMvRxBundle
+import im.vector.app.core.platform.VectorBaseActivity
+import im.vector.app.core.utils.toast
+import im.vector.app.databinding.ActivitySimpleBinding
+import im.vector.app.features.home.room.detail.upgrade.MigrateRoomBottomSheet
+import im.vector.app.features.roomprofile.RoomProfileArgs
+import im.vector.app.features.roomprofile.settings.joinrule.advanced.RoomJoinRuleChooseRestrictedActions
+import im.vector.app.features.roomprofile.settings.joinrule.advanced.RoomJoinRuleChooseRestrictedEvents
+import im.vector.app.features.roomprofile.settings.joinrule.advanced.RoomJoinRuleChooseRestrictedState
+import im.vector.app.features.roomprofile.settings.joinrule.advanced.RoomJoinRuleChooseRestrictedViewModel
+import javax.inject.Inject
+
+class RoomJoinRuleActivity : VectorBaseActivity<ActivitySimpleBinding>(),
+        RoomJoinRuleChooseRestrictedViewModel.Factory {
+
+    override fun getBinding() = ActivitySimpleBinding.inflate(layoutInflater)
+
+    private lateinit var roomProfileArgs: RoomProfileArgs
+
+    @Inject
+    lateinit var allowListViewModelFactory: RoomJoinRuleChooseRestrictedViewModel.Factory
+
+    @Inject
+    lateinit var errorFormatter: ErrorFormatter
+
+    val viewModel: RoomJoinRuleChooseRestrictedViewModel by viewModel()
+
+    override fun create(initialState: RoomJoinRuleChooseRestrictedState) = allowListViewModelFactory.create(initialState)
+
+    override fun injectWith(injector: ScreenComponent) {
+        injector.inject(this)
+    }
+
+    override fun initUiAndData() {
+        roomProfileArgs = intent?.extras?.getParcelable(MvRx.KEY_ARG) ?: return
+        if (isFirstCreation()) {
+            addFragment(
+                    R.id.simpleFragmentContainer,
+                    RoomJoinRuleFragment::class.java,
+                    roomProfileArgs
+            )
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel.selectSubscribe(this, RoomJoinRuleChooseRestrictedState::updatingStatus) {
+            when (it) {
+                Uninitialized -> {
+                    // nop
+                }
+                is Loading    -> {
+                    views.simpleActivityWaitingView.isVisible = true
+                }
+                is Success    -> {
+                    withState(viewModel) { state ->
+                        if (state.didSwitchToReplacementRoom) {
+                            // we should navigate to new room
+                            navigator.openRoom(this, state.roomId, null, true)
+                        }
+                        finish()
+                    }
+                }
+                is Fail       -> {
+                    views.simpleActivityWaitingView.isVisible = false
+                    toast(errorFormatter.toHumanReadable(it.error))
```

Comment:
You should use ViewEvents to display errors, to avoid displaying it multiple times.